storage/database/sharded_database.go

```diff
@@ -151,102 +154,248 @@ func (pdb *shardedDB) Close() {
+}
+
+// NewIteratorWithPrefixUnsorted creates a iterator over a subset of database content
+// with a particular key prefix.
+func (db *shardedDB) NewIteratorWithPrefixUnsorted(prefix []byte) Iterator {
+	return db.newIteratorUnsorted(func(db Database) Iterator { return db.NewIteratorWithPrefix(prefix) })
+}
+
+func (db *shardedDB) newIteratorUnsorted(newIterator func(Database) Iterator) Iterator {
+	resultCh := make(chan entry, shardedDBCombineChanSize)
+	it := &shardedDBIteratorUnsorted{
+		shardedDBIterator{
+			db.NewChanIterator(context.Background(), resultCh, newIterator),
+			resultCh,
+			nil, nil}}
+	return it
+}
+
+// shardedDBChanIterator creates iterators for each shard DB.
+// Channels subscribing each iterators can be gained.
+// Each iterators fetch values in binary-alphabetical order.
+// This is useful when you want to operate on each items in parallel.
+type shardedDBChanIterator struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	iterators []Iterator
+
+	combinedChan bool // all workers put items to one resultChan
+	shardNum     int  // num of shards left to iterate
+	resultChs    []chan entry
+}
+
+// NewChanIterator creates iterators for each shard DB. This is useful when you
+// want to operate on each items in parallel.
+// If `resultCh` is given, all items are written to `resultCh`, unsorted. If
+// `resultCh` is not given, new channels are created for each DB. Items are
+// written to corresponding channels in binary-alphabetical order. The channels
+// can be gained by calling `Channels()`.
+//
+// If you want to get ordered items in serial, checkout shardedDB.NewIterator()
+// If you want to get unordered items in serial with Iterator Interface,
+// checkout shardedDB.NewIteratorUnsorted().
+func (db *shardedDB) NewChanIterator(ctx context.Context, resultCh chan entry, newIterator func(Database) Iterator) shardedDBChanIterator {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	it := shardedDBChanIterator{ctx, nil,
+		make([]Iterator, len(db.shards)),
+		resultCh != nil,
+		len(db.shards),
+		make([]chan entry, len(db.shards))}
+	it.ctx, it.cancel = context.WithCancel(ctx)
+
+	for i := 0; i < len(db.shards); i++ {
+		it.iterators[i] = newIterator(db.shards[i])
+		if resultCh == nil {
+			it.resultChs[i] = make(chan entry, shardedDBSubChannelSize)
+		} else {
+			it.resultChs[i] = resultCh
+		}
+		go it.runChanWorker(it.iterators[i], it.resultChs[i], it.ctx)
+	}
+
+	return it
+}
+
+func (sit *shardedDBChanIterator) runChanWorker(it Iterator, resultCh chan entry, ctx context.Context) {
+iter:
+	for it.Next() {
+		select {
+		case <-ctx.Done():
+			break iter
+		default:
+		}
+		key := make([]byte, len(it.Key()))
+		val := make([]byte, len(it.Value()))
+		copy(key, it.Key())
+		copy(val, it.Value())
+		resultCh <- entry{key, val}
+	}
+	it.Release()
+	if sit.shardNum--; sit.combinedChan && sit.shardNum > 0 {
+		return
+	}
+	close(resultCh)
```

Comment:
Could you explain why you think that way?
Each channel worker closes its own iterator. Therefore `it.Release()` should be called allways.
If the worker is using `combinedChan`, only the the last worker should close the channel. If not,  all workers always closes their channel, like an iterator release.

Reply:
I asked since running a worker serves both jobs, retrieving value from iterator and releasing the related resources. However, I understood your point. Thanks for your explanation.